src/Microsoft.Extensions.Configuration/ConfigurationBuilder.cs

```diff
@@ -30,41 +31,37 @@ public IEnumerable<IConfigurationProvider> Providers
         public Dictionary<string, object> Properties { get; } = new Dictionary<string, object>();
 
         /// <summary>
-        /// Adds a new configuration provider.
+        /// Adds a new configuration source.
         /// </summary>
-        /// <param name="provider">The configuration provider to add.</param>
+        /// <param name="source">The configuration source to add.</param>
         /// <returns>The same <see cref="IConfigurationBuilder"/>.</returns>
-        public IConfigurationBuilder Add(IConfigurationProvider provider)
+        public IConfigurationBuilder Add(IConfigurationSource source)
         {
-            return Add(provider, load: true);
-        }
-
-        /// <summary>
-        /// Adds a new provider to obtain configuration values from.
-        /// This method is intended only for test scenarios.
-        /// </summary>
-        /// <param name="provider">The configuration provider to add.</param>
-        /// <param name="load">If true, the configuration provider's <see cref="IConfigurationProvider.Load"/> method will
-        ///  be called.</param>
-        /// <returns>The same <see cref="IConfigurationBuilder"/>.</returns>
-        public IConfigurationBuilder Add(IConfigurationProvider provider, bool load)
-        {
-            if (load)
+            if (source == null)
             {
-                provider.Load();
+                throw new ArgumentNullException(nameof(source));
             }
-            _providers.Add(provider);
+
+            _sources.Add(source);
             return this;
         }
 
         /// <summary>
         /// Builds an <see cref="IConfiguration"/> with keys and values from the set of providers registered in
-        /// <see cref="Providers"/>.
+        /// <see cref="Sources"/>.
         /// </summary>
         /// <returns>An <see cref="IConfigurationRoot"/> with keys and values from the registered providers.</returns>
         public IConfigurationRoot Build()
         {
-            return new ConfigurationRoot(_providers);
+            var providers = new List<IConfigurationProvider>();
+            foreach (var source in _sources)
+            {
+                var provider = source.Build(this);
+                // REVIEW: who should call load?  source.Build?
```

Comment:
I think the builder should actually call Load, I tried it via the provider but it just seems like that makes it easy to forget to call Load and have weird behavior.

Feels better to just have the Builder do it before Add()...

```
                var provider = source.Build(this);
                provider.Load();
                providers.Add(provider);
```